File: 2024/04/part1.py
import fileinput


def read_input():
    for line in fileinput.input():
        yield line.strip()


def scan_horizontal(width, height):
    for y in range(height):
        yield ((x, y) for x in range(width))


def scan_vertical(width, height):
    for x in range(width):
        yield ((x, y) for y in range(height))


def scan_diag_backslash(width, height):
    for x in range(width):
        yield ((x + d, d) for d in range(min(width - x, height)))
    for y in range(1, height):
        yield ((d, y + d) for d in range(min(width, height - y)))


def scan_diag_slash(width, height):
    for x in range(width):
        yield ((x + d, height - 1 - d) for d in range(min(width - x, height)))
    for y in range(height - 1):
        yield ((d, y - d) for d in range(min(width, y + 1)))


def bidirectional(scanner):
    for scan in scanner:
        coords = list(scan)
        yield iter(coords)
        yield reversed(coords)


def scan_grid(width, height):
    yield from bidirectional(scan_horizontal(width, height))
    yield from bidirectional(scan_vertical(width, height))
    yield from bidirectional(scan_diag_backslash(width, height))
    yield from bidirectional(scan_diag_slash(width, height))


def solve(data):
    height = len(data)
    width = len(data[0])
    count = 0
    # for coords in scan_grid(width, height):
    #     print(" ".join(repr(coord) for coord in coords))
    for coords in scan_grid(width, height):
        line = "".join(data[y][x] for x, y in coords)
        count += line.count("XMAS")
    print(count)


if __name__ == "__main__":
    solve(tuple(read_input()))

File: 2024/04/part1.rs
use std::{cmp, env, fs::File, io, io::BufRead, path::Path};

fn read_input(path: &Path) -> io::Result<Vec<String>> {
    let file = File::open(path)?;
    let mut data: Vec<String> = Vec::new();
    for line in io::BufReader::new(file).lines() {
        data.push(line?);
    }
    Ok(data)
}

#[derive(Clone, Debug)]
struct GridScan {
    start: (usize, usize),
    end: (usize, usize),
    step: (isize, isize),
}

impl Iterator for GridScan {
    type Item = (usize, usize);

    fn next(&mut self) -> Option<Self::Item> {
        let curr = self.start;
        if curr == self.end {
            return None
        }
        let (mut x, mut y) = curr;
        let (dx, dy) = self.step;
        x = x.wrapping_add_signed(dx);
        y = y.wrapping_add_signed(dy);
        self.start = (x, y);
        Some(curr)
    }
}

impl DoubleEndedIterator for GridScan {
    fn next_back(&mut self) -> Option<Self::Item> {
        let curr = self.end;
        if curr == self.start {
            return None
        }
        let (mut x, mut y) = curr;
        let (dx, dy) = self.step;
        x = x.wrapping_add_signed(-dx);
        y = y.wrapping_add_signed(-dy);
        self.end = (x, y);
        Some(self.end)
    }
}

impl GridScan {
    fn horizontal(width: usize, height: usize) -> impl Iterator<Item=GridScan> {
        (0..height).map(move |y|
            GridScan { start: (0, y), end: (width, y), step: (1, 0) }
        )
    }

    fn vertical(width: usize, height: usize) -> impl Iterator<Item=GridScan> {
        (0..width).map(move |x|
            GridScan { start: (x, 0), end: (x, height), step: (0, 1) }
        )
    }

    fn backslash(width: usize, height: usize) -> impl Iterator<Item=GridScan> {
        (0..width).map(move |x| {
            let num_steps = cmp::min(width - x, height);
            GridScan { start: (x, 0), end: (x + num_steps, num_steps), step: (1, 1) }
        }).chain((1..height).map(move |y| {
            let num_steps = cmp::min(width, height - y);
            GridScan { start: (0, y), end: (num_steps, y + num_steps), step: (1, 1) }
        }))
    }

    fn slash(width: usize, height: usize) -> impl Iterator<Item=GridScan> {
        (0..width).map(move |x| {
            let num_steps = cmp::min(width - x, height);
            GridScan { start: (x, height - 1), end: (x + num_steps, (height - 1).wrapping_sub(num_steps)), step: (1, -1) }
        }).chain((0..height - 1).map(move |y| {
            let num_steps = cmp::min(width, y + 1);
            GridScan { start: (0, y), end: (num_steps, y.wrapping_sub(num_steps)), step: (1, -1) }
        }))
    }
}

fn bidirectional(scanners: impl Iterator<Item=GridScan>) -> impl Iterator<Item=Box<dyn Iterator<Item=(usize, usize)>>> {
    scanners.flat_map(|scanner| {
        let dirs: [Box<dyn Iterator<Item=(usize, usize)>>; 2] = [
            Box::new(scanner.clone()),
            Box::new(scanner.rev()),
        ];
        dirs.into_iter()
    })
}

const SCANNERS: [fn(usize, usize) -> Box<dyn Iterator<Item=GridScan>>; 4] = [
    |w, h| Box::new(GridScan::horizontal(w, h)),
    |w, h| Box::new(GridScan::vertical(w, h)),
    |w, h| Box::new(GridScan::backslash(w, h)),
    |w, h| Box::new(GridScan::slash(w, h)),
];

fn scan_grid(width: usize, height: usize) -> impl Iterator<Item=Box<dyn Iterator<Item=(usize, usize)>>> {
    SCANNERS.iter().flat_map(move |scanner|
        bidirectional(scanner(width, height))
    )
}

fn solve(grid: &Vec<String>) {
    let height = grid.len();
    let width = grid[0].len();
    let count: usize = scan_grid(width, height).into_iter().map(|scanner| {
        let line: String = scanner.map(|(x, y)| grid[y].chars().nth(x).unwrap()).collect();
        line.matches("XMAS").count()
    }).sum();
    println!("{count}");
}

fn main() {
    for arg in env::args().skip(1) {
        println!("=== {arg}");
        let data = read_input(Path::new(&arg)).expect("reading input");
        solve(&data);
    }
}
